src/TypePredicates.hs

```diff
@@ -0,0 +1,30 @@
+module TypePredicates where
+
+import Types
+
+isTypeGeneric :: Ty -> Bool
+isTypeGeneric (VarTy _) = True
+isTypeGeneric (FuncTy argTys retTy _) = any isTypeGeneric argTys || isTypeGeneric retTy
+isTypeGeneric (StructTy n tyArgs) = isTypeGeneric n || any isTypeGeneric tyArgs
+isTypeGeneric (PointerTy p) = isTypeGeneric p
+isTypeGeneric (RefTy r _) = isTypeGeneric r
+isTypeGeneric _ = False
+
+isFullyGenericType :: Ty -> Bool
+isFullyGenericType (VarTy _) = True
```

Comment:
We actually need to add a case here:

`(StructTy name members) = isFullyGenericType name && all isFullyGenericType members`

Since structs can now have variable names like `(f a)`

Reply:
Alright, adding it!